File: manji-standard-server-ts-next/src/domain/service/user-service.ts
import { randomUUID } from "node:crypto";
import { User } from "@/domain/entity/user";
import type { UserRepository } from "@/domain/repository/user-repository";

export class EmailAlreadyTakenError extends Error {
  constructor() {
    super("email already taken");
    this.name = "EmailAlreadyTakenError";
  }
}

export class UserService {
  constructor(
    private readonly userRepo: UserRepository,
    private readonly clock: () => Date = () => new Date(),
  ) {}

  async create(email: string, name: string): Promise<User> {
    const existing = await this.userRepo.findByEmail(email);
    if (existing) throw new EmailAlreadyTakenError();
    const user = User.create({
      id: randomUUID(),
      email,
      name,
      createdAt: this.clock(),
    });
    await this.userRepo.save(user);
    return user;
  }

  async getById(id: string): Promise<User | null> {
    return this.userRepo.findById(id);
  }
}

File: manji-standard-server-go/pkg/domain/service/user_service.go
package service

import (
	"context"
	"errors"
	"time"

	"github.com/example/manji-standard-server-go/pkg/domain/entity"
	"github.com/example/manji-standard-server-go/pkg/domain/repository"
	"github.com/google/uuid"
)

type UserService struct {
	userRepo repository.UserRepository
	clock    func() time.Time
}

func NewUserService(userRepo repository.UserRepository, clock func() time.Time) *UserService {
	if clock == nil {
		clock = time.Now
	}
	return &UserService{userRepo: userRepo, clock: clock}
}

var ErrEmailAlreadyTaken = errors.New("email already taken")

func (s *UserService) Create(ctx context.Context, email, name string) (*entity.User, error) {
	existing, err := s.userRepo.SelectByEmail(ctx, email)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return nil, ErrEmailAlreadyTaken
	}

	user, err := entity.NewUser(uuid.NewString(), email, name, s.clock())
	if err != nil {
		return nil, err
	}
	if err := s.userRepo.Insert(ctx, user); err != nil {
		return nil, err
	}
	return user, nil
}

func (s *UserService) GetByID(ctx context.Context, id string) (*entity.User, error) {
	return s.userRepo.SelectByPK(ctx, id)
}
